File: worker/src/services/reranker.ts
import { TanaNode, ScoredCandidate } from '../types';
import { computeTemporalScore, classifyTemporalRelationship, isDayNode } from './temporal';

export class EdgeReranker {
  /**
   * Multi-factor reranking based on semantic score, temporal proximity/provenance, structural depth, and deduplication.
   */
  rerank(
    task: string,
    candidates: Array<{ node: TanaNode; score: number; reason: string }>,
    maxNodes: number = 8,
    targetDate?: string,
    dateFrom?: string,
    dateTo?: string,
    temporalMode: string = 'none'
  ): ScoredCandidate[] {
    const taskTokens = new Set((task.toLowerCase().match(/\w+/g) || []).filter(t => t.length > 2));
    const now = Date.now();
    const targetYear = targetDate ? targetDate.substring(0, 4) : (dateFrom ? dateFrom.substring(0, 4) : undefined);

    const scored: ScoredCandidate[] = [];

    for (const c of candidates) {
      let finalScore = c.score;
      const cleanName = c.node.name.trim();

      // Penalize placeholder/empty template titles
      if ((cleanName.toLowerCase() === 'title:' || cleanName.toLowerCase() === 'untitled' || cleanName.length < 3) && (!c.node.description || c.node.description.length < 10)) {
        finalScore -= 0.35;
      }

      // 1. Exact Lexical Overlap Boost
      const nameAndDesc = `${cleanName} ${c.node.description || ''}`.toLowerCase();
      const nodeTokens = new Set((nameAndDesc.match(/\w+/g) || []));
      let nameOverlapCount = 0;
      for (const t of taskTokens) {
        if (nodeTokens.has(t)) nameOverlapCount++;
      }
      if (taskTokens.size > 0 && nameOverlapCount > 0) {
        const overlapRatio = nameOverlapCount / taskTokens.size;
        finalScore += Math.min(0.40, overlapRatio * 0.45);
      }

      // 2. High-Priority Field Name & Value Scoring
      let fieldMatchCount = 0;
      const matchedFieldDetails: string[] = [];
      const hasFields = c.node.fields && c.node.fields.length > 0;

      if (hasFields) {
        for (const f of c.node.fields) {
          const fName = (f.field_name || '').toLowerCase();
          const fVal = (f.value_text || f.value_node_id || '').toLowerCase();
          const fieldTokens = new Set(`${fName} ${fVal}`.match(/\w+/g) || []);

          let fieldHits = 0;
          for (const t of taskTokens) {
            if (fieldTokens.has(t)) fieldHits++;
          }

          if (fieldHits > 0) {
            fieldMatchCount += fieldHits;
            matchedFieldDetails.push(`${f.field_name}: ${f.value_text || f.value_node_id}`);
          }
        }

        if (taskTokens.size > 0 && fieldMatchCount > 0) {
          finalScore += Math.min(0.30, (fieldMatchCount / taskTokens.size) * 0.35);
        }
        finalScore += Math.min(0.08, c.node.fields.length * 0.02);
      }

      // 3. Temporal Handling
      if (targetDate || ['strict', 'boost', 'filter'].includes(temporalMode)) {
        const tempScore = computeTemporalScore(c.node.effective_date, targetDate, dateFrom, dateTo);
        const relClass = classifyTemporalRelationship(c.node, targetDate, targetYear);

        if (temporalMode === 'strict') {
          if (relClass === 'contemporaneous') {
            finalScore += 0.30 + (tempScore * 0.20);
          } else if (relClass === 'retrospective') {
            finalScore -= 0.50;
          } else if (relClass === 'conflicting') {
            finalScore -= 0.60;
          } else {
            finalScore -= 0.40;
          }
        } else if (temporalMode === 'boost') {
          finalScore = (finalScore * 0.60) + (tempScore * 0.40);
        }
      } else {
        // General query recency boost
        if (c.node.updated_at) {
          try {
            const updatedTime = new Date(c.node.updated_at).getTime();
            const ageDays = (now - updatedTime) / (1000 * 60 * 60 * 24);
            if (ageDays < 30) {
              finalScore += 0.05 * (1.0 - ageDays / 30);
            }
          } catch {
            // Ignore parse errors
          }
        }
      }

      // 4. Supertag relevance
      if (c.node.supertags && c.node.supertags.length > 0) {
        finalScore += 0.05;
      }

      // 5. Bare calendar container demotion
      const [isBareCal] = isDayNode(c.node);
      if (isBareCal && !c.node.description && !hasFields) {
        finalScore *= 0.40;
      }

      let annotatedReason = c.reason;
      if (matchedFieldDetails.length > 0) {
        annotatedReason = `${c.reason} (Matched field: ${matchedFieldDetails.slice(0, 2).join(', ')})`;
      } else if (nameOverlapCount === taskTokens.size && taskTokens.size > 0) {
        annotatedReason = `High-confidence title & content match for '${task.slice(0, 40)}'`;
      }

      scored.push({
        node: c.node,
        score: finalScore,
        reason: annotatedReason
      });
    }

    // Sort descending by score
    scored.sort((a, b) => b.score - a.score);

    // Context deduplication
    const seenSignatures = new Set<string>();
    const deduplicated: ScoredCandidate[] = [];

    for (const item of scored) {
      const sig = `${(item.node.parent_id || '')}::${item.node.name.trim().toLowerCase()}`;
      if (!seenSignatures.has(sig)) {
        seenSignatures.add(sig);
        deduplicated.push(item);
      }
      if (deduplicated.length >= maxNodes) break;
    }

    return deduplicated;
  }
}

File: tana_context_mask/engine/reranker.py
import math
from datetime import datetime
from typing import List, Dict, Tuple, Any, Optional
from ..models.node import TanaNode
from .temporal import compute_temporal_score, classify_temporal_relationship, is_day_node

class TaskReranker:
    def __init__(self):
        pass

    def rerank(
        self,
        task_query: str,
        candidates: List[Tuple[TanaNode, float, str]],  # (node, base_score, inclusion_reason)
        max_nodes: int = 8,
        target_date: Optional[str] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None,
        temporal_mode: str = "none"
    ) -> List[Tuple[TanaNode, float, str]]:
        """
        Reranks expanded node candidates according to multi-factor relevance:
        - Base semantic/lexical similarity
        - Graph centrality and structural connectivity
        - Temporal proximity (when target_date / history mode active) or recency decay (general queries)
        - Tag importance and content completeness
        """
        if not candidates:
            return []

        all_candidate_ids = {node.id for node, _, _ in candidates}
        scored_candidates: List[Tuple[TanaNode, float, str]] = []

        now = datetime.now().timestamp()
        target_year = target_date[:4] if target_date else (date_from[:4] if date_from else None)

        for node, base_score, reason in candidates:
            score = base_score

            # 1. Graph Connectivity Bonus
            cluster_connections = 0
            if node.parent_id in all_candidate_ids:
                cluster_connections += 1
            for cid in node.children:
                if cid in all_candidate_ids:
                    cluster_connections += 1
            for rid in node.references:
                if rid in all_candidate_ids:
                    cluster_connections += 1

            if cluster_connections > 0:
                score += min(0.25, cluster_connections * 0.08)

            # 2. Temporal Handling
            if target_date or (temporal_mode in ["strict", "boost", "filter"]):
                temp_score = compute_temporal_score(
                    node_date_str=node.effective_date,
                    target_date_str=target_date,
                    date_from_str=date_from,
                    date_to_str=date_to
                )
                rel_class = classify_temporal_relationship(
                    node=node,
                    target_date_str=target_date,
                    target_year=target_year
                )

                if temporal_mode == "strict":
                    if rel_class == "contemporaneous":
                        score += 0.30 + (temp_score * 0.20)
                    elif rel_class == "retrospective":
                        score -= 0.50  # Strong demotion so contemporary evidence dominates
                    elif rel_class == "conflicting":
                        score -= 0.60
                    else:
                        score -= 0.40
                elif temporal_mode == "boost":
                    score = (score * 0.60) + (temp_score * 0.40)
            else:
                # General query: Recent updates get small bonus
                if node.updated_at:
                    try:
                        updated_ts = datetime.fromisoformat(node.updated_at.replace("Z", "+00:00")).timestamp()
                        days_old = max(0, (now - updated_ts) / 86400.0)
                        recency_boost = 0.10 * math.exp(-days_old / 180.0)
                        score += recency_boost
                    except Exception:
                        pass

            # 3. Structural Significance Bonus (Supertags)
            tag_names = [t.tag_name.lower() for t in node.supertags]
            if any(t in tag_names for t in ["project", "goal", "task", "milestone", "meeting", "sprint"]):
                score += 0.08

            # 4. Bare calendar container demotion & penalty for trivial nodes
            is_bare_cal, _ = is_day_node(node)
            if is_bare_cal and not node.description and not node.fields:
                score *= 0.40
            elif len(node.name.strip()) < 4 and not node.description and not node.fields:
                score *= 0.60

            scored_candidates.append((node, round(score, 4), reason))

        # Sort descending by composite score
        scored_candidates.sort(key=lambda x: x[1], reverse=True)

        # Deduplication
        seen_keys = set()
        deduped: List[Tuple[TanaNode, float, str]] = []
        for node, s, r in scored_candidates:
            norm_key = (node.parent_id or "", node.name.strip().lower())
            if norm_key in seen_keys and len(node.name.strip()) > 10:
                continue
            seen_keys.add(norm_key)
            deduped.append((node, s, r))
            if len(deduped) >= max_nodes:
                break

        return deduped
